### PR TITLE
Bug 1968572: Assisted Service does not escape backslash characters on public SSH keys

### PR DESCRIPTION
**Pull request by sagidayan**
# Description
 - Change function `getUserSSHKey` from a string based to a marshaled struct based
 - Function will return a stringified JSON block and an error if needed
 - In `FormatDiscoveryIgnitionFile` will return an error if `getUserSSHKey` returns an error

# What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

# How was this code tested?

Please, select one or more if needed:

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

It's recommended to take a few extra minutes to provide more information about
how this code was tested. Here are some questions that may be worth answering:

- Should this PR be tested by the reviewer?
- Is this PR relying on CI for an e2e test run?
- Should this PR be tested in a specific environment?
- Any logs, screenshots, etc that can help with the review process?

# Assignees

Please, add one or two reviewers that could help review this PR. Use `/assign` if you want to assign
this PR directly to someone.

/cc @
/cc @

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/ignition/ignition.go [hunk 1]
```diff
@@ -1271,8 +1271,12 @@ func (ib *ignitionBuilder) FormatDiscoveryIgnitionFile(cluster *common.Cluster,
 	if cfg.InstallRHCa {
 		rhCa = url.PathEscape(RedhatRootCA)
 	}
+	userSshKey, err := getUserSSHKey(cluster.ImageInfo.SSHPublicKey)
+	if err != nil {
+		return "", err
```
Comment (michaellevy101) [inline internal/ignition/ignition.go:1276]: please wrap this message and log it

File: internal/ignition/ignition.go [hunk 2]
```diff
@@ -1388,16 +1392,33 @@ func QuoteSshPublicKeys(sshPublicKeys string) string {
...
 
-func getUserSSHKey(sshKey string) string {
+func getUserSSHKey(sshKey string) (string, error) {
 	if sshKey == "" {
-		return ""
+		return "", nil
+	}
+	keys := strings.Split(sshKey, "\n")
+	validKeys := []string{}
+	// filter only valid non empty keys
```
Comment (michaellevy101) [inline internal/ignition/ignition.go:1401]: Please consider fix the input when receiving it from the customer

File: internal/ignition/ignition_test.go [hunk 3]
```diff
@@ -1072,3 +1072,83 @@ var _ = Describe("IgnitionBuilder", func() {
 		})
 	})
 })
+
+var _ = Describe("Ignition SSH key escaping", func() {
```
Comment (michaellevy101) [inline internal/ignition/ignition_test.go:1076]: Ignition SSH key building

File: internal/ignition/ignition_test.go [hunk 4]
```diff
@@ -1072,3 +1072,83 @@ var _ = Describe("IgnitionBuilder", func() {
...
+		cluster = common.Cluster{Cluster: models.Cluster{
+			ID:            &clusterID,
+			PullSecretSet: false,
+		}, PullSecret: "{\"auths\":{\"cloud.openshift.com\":{\"auth\":\"dG9rZW46dGVzdAo=\",\"email\":\"coyote@acme.com\"}}}"}
+		cluster.ImageInfo = &models.ImageInfo{}
+		builder = NewBuilder(log, mockStaticNetworkConfig, mockMirrorRegistriesConfigBuilder)
+	})
+	Context("when empty or invalid input", func() {
+		It("white_space_string should return an empty string", func() {
+			mockMirrorRegistriesConfigBuilder.EXPECT().IsMirrorRegistriesConfigured().Return(false).Times(1)
```
Comment (michaellevy101) [inline internal/ignition/ignition_test.go:1099]: extract to BeforeEach

File: internal/ignition/ignition_test.go [hunk 5]
```diff
@@ -1072,3 +1072,83 @@ var _ = Describe("IgnitionBuilder", func() {
...
+			cluster.ImageInfo.SSHPublicKey = "  \n  \n \t \n  "
+			text, err := builder.FormatDiscoveryIgnitionFile(&cluster, IgnitionConfig{}, false, auth.TypeRHSSO)
```
Comment (michaellevy101) [inline internal/ignition/ignition_test.go:1101]: extract to function

**Review by michaellevy101 — COMMENTED** [review]
Well done!!

File: internal/ignition/ignition_test.go (hunk 3 above)
Comment (sagidayan) [reply internal/ignition/ignition_test.go:1076]: 👍🏽

File: internal/ignition/ignition_test.go (hunk 4 above)
Comment (sagidayan) [reply internal/ignition/ignition_test.go:1099]: 👍🏽

File: internal/ignition/ignition_test.go (hunk 5 above)
Comment (sagidayan) [reply internal/ignition/ignition_test.go:1101]: 👍🏽

File: internal/ignition/ignition.go (hunk 1 above)
Comment (sagidayan) [reply internal/ignition/ignition.go:1277]: 👍🏽

File: internal/ignition/ignition.go (hunk 2 above)
Comment (sagidayan) [reply internal/ignition/ignition.go:1402]: The thing is that if we already have some values in DB like `ssh-rsa key coyote@acme.com\n\n\tssh-rsa key2 coyote@acme.com` Then we might just want to filter out the empty strings out of the keys array to make it cleaner and more readable. 
But that's true, we should also "fix" it before DB insertion.

**Comment by romfreiman** [comment]
Do we know how test it in the e2e flow? Not with ut

**Pull request merged**